frontend/openchat-agent/src/services/openchatAgent.ts

```diff
@@ -851,6 +871,164 @@ export class OpenChatAgent extends EventTarget {
+        const groupChatIds = current.groupChats.map((g) => g.chatId).concat(userResponse.groupChatsAdded.map((g) => g.chatId));
+        const groupIndexResponse = await this._groupIndexClient.filterGroups(groupChatIds, current.timestamp);
+
+        const activeGroups = new Set(groupIndexResponse.activeGroups);
+
+        const groupPromises = userResponse.groupChatsAdded.map((g) => this.getGroupClient(g.chatId).summary());
+        const groupUpdatePromises = current.groupChats
+            .filter((g) => activeGroups.has(g.chatId))
+            .map((g) => this.getGroupClient(g.chatId).summaryUpdates(g.lastUpdated));
+
+        const groupPromiseResults = await waitAll(groupPromises);
+        const groupUpdatePromiseResults = await waitAll(groupUpdatePromises);
+
+        const groups = groupPromiseResults.success.filter(isSuccessfulGroupSummaryResponse);
+        const groupUpdates = groupUpdatePromiseResults.success.filter(isSuccessfulGroupSummaryUpdatesResponse);
+
+        const anyUpdates =
+            userResponse.directChatsAdded.length > 0 ||
+            userResponse.directChatsUpdated.length > 0 ||
+            userResponse.groupChatsAdded.length > 0 ||
+            userResponse.groupChatsUpdated.length > 0 ||
+            userResponse.chatsRemoved.length > 0 ||
+            userResponse.avatarId !== undefined ||
+            userResponse.blockedUsers !== undefined ||
+            userResponse.pinnedChats !== undefined ||
+            groups.length > 0 ||
+            groupUpdates.length > 0;
+
+        const anyErrors =
+            groupPromiseResults.errors.length > 0 ||
+            groupUpdatePromiseResults.errors.length > 0;
+
+        const directChats = userResponse.directChatsAdded.concat(
+            mergeDirectChatUpdates(current.directChats, userResponse.directChatsUpdated));
+
+        const chatsRemoved = new Set(userResponse.chatsRemoved.concat(groupIndexResponse.deletedGroups.map((g) => g.id)));
+
+        const groupChats = mergeGroupChats(userResponse.groupChatsAdded, groups)
+            .concat(mergeGroupChatUpdates(current.groupChats, userResponse.groupChatsUpdated, groupUpdates))
+            .filter((g) => !chatsRemoved.has(g.chatId));
+
+        const state = {
+            timestamp: userResponse.timestamp,
+            directChats,
+            groupChats,
+            avatarId: applyOptionUpdate(current.avatarId, userResponse.avatarId),
+            blockedUsers: userResponse.blockedUsers ?? current.blockedUsers,
+            pinnedChats: userResponse.pinnedChats ?? current.pinnedChats,
+        }
+        const affectedEvents = getAffectedEvents(userResponse.directChatsUpdated, groupUpdates);
+
+        await setCachedChatsV2(this.db, this.principal, state);
+
+        return await this.hydrateChatState(state).then((s) => ({
+            state: s,
+            affectedEvents,
+            anyUpdates,
+            anyErrors,
+        }));
+    }
+
```

Comment:
seems like we might want to use `waitAll` here as well

Reply:
I only used `waitAll` for groups because their updates are processed individually, so if a group fails, it can simply be tried again on the next iteration.
Whereas if hydrating a direct chat fails, we can't return all of the other direct chats and skip the one that failed, because there is only a single timestamp for all direct chats which means we could miss updates for the one that failed because the timestamp will have been incremented.
I think we can improve on this in the future though to make it more resilient.